src/main/java/spoon/support/visitor/SubInheritanceHierarchyResolver.java

```diff
@@ -0,0 +1,232 @@
+		this.includingInterfaces = includingInterfaces;
+		return this;
+	}
+
+	/**
+	 * @param failOnClassNotFound sets whether processing should throw an exception if class is missing in noClassPath mode
+	 */
+	public SubInheritanceHierarchyResolver failOnClassNotFound(boolean failOnClassNotFound) {
+		this.failOnClassNotFound = failOnClassNotFound;
+		return this;
+	}
+
+	/**
+	 * Calls `outputConsumer.apply(subType)` for each sub type of the targetSuperTypes that are found in `inputPackage`.
+	 * Each sub type is returned only once.
+	 * It makes sense to call this method again for example after new super types are added
+	 * by {@link #addSuperType(CtTypeInformation)}.
+	 *
+	 * 	If this method is called again with same input and configuration, nothing in sent to outputConsumer
+	 * @param outputConsumer the consumer for found sub types
+	 */
+	public <T extends CtTypeInformation> void forEachSubTypeInPackage(final CtConsumer<T> outputConsumer) {
+		/*
+		 * Set of qualified names of all visited types, independent on whether they are sub types or not.
+		 */
+		final Set<String> allVisitedTypeNames = new HashSet<>();
+		/*
+		 * the queue of types whose super inheritance hierarchy we are just visiting.
+		 * They are potential sub types of an `targetSuperTypes`
+		 */
+		final Deque<CtTypeReference<?>> currentSubTypes = new ArrayDeque<>();
+		//algorithm
+		//1) query step: scan input package for sub classes and sub interfaces
+		final CtQuery q = inputPackage.map(new CtScannerFunction());
+		//2) query step: visit only required CtTypes
+		if (includingInterfaces) {
+			//the client is interested in sub inheritance hierarchy of interfaces too. Check interfaces, classes, enums, Annotations, but not CtTypeParameters.
+			q.select(typeFilter);
+		} else {
+			//the client is not interested in sub inheritance hierarchy of interfaces. Check only classes and enums.
+			q.select(classFilter);
+		}
+		/*
+		 * 3) query step: for each found CtType, visit it's super inheritance hierarchy and search there for a type which is equal to one of targetSuperTypes.
+		 * If found then all sub types in hierarchy (variable `currentSubTypes`) are sub types of targetSuperTypes. So return them
+		 */
+		q.map(new SuperInheritanceHierarchyFunction()
+			//if there is any interface between `targetSuperTypes`, then we have to check superInterfaces too
+			.includingInterfaces(hasSuperInterface)
+			.failOnClassNotFound(failOnClassNotFound)
+			/*
+			 * listen for types in super inheritance hierarchy
+			 * 1) to collect `currentSubTypes`
+			 * 2) to check if we have already found a targetSuperType
+			 * 3) if found then send `currentSubTypes` to `outputConsumer` and skip visiting of further super types
+			 */
+			.setListener(new CtScannerListener() {
+				@Override
+				public ScanningMode enter(CtElement element) {
+					final CtTypeReference<?> typeRef = (CtTypeReference<?>) element;

```

Comment:
This case is always right because `SuperInheritanceHierarchyFunction` by default always returns TypeReferences. My origin code declared it explicitly on line 145 where was
```java
//internally it works always with references
.returnTypeReferences(true)
```
but Martin cleared that ;-)